browserdebuggertools/wssessionmanager.py

```diff
@@ -15,23 +17,116 @@
+        self._continue = True
+        self._poll_signal = poll_signal
+
+        self.exception = None
+        self.ws = self._get_websocket()
+        self.daemon = True
+
+    def __del__(self):
+        self.close()
+
+    def _get_websocket_url(self, port):
+        response = requests.get(
+            "http://localhost:{}/json".format(port), timeout=self._CONN_TIMEOUT
+        )
+        if not response.ok:
+            raise DevToolsException("{} {} for url: {}".format(
+                response.status_code, response.reason, response.url)
+            )
+
+        tabs = [target for target in response.json() if target["type"] == "page"]
+        if not tabs:
+            raise TabNotFoundError("There is no tab to connect to.")
+        return tabs[0]["webSocketDebuggerUrl"]
+
+    def _get_websocket(self):
+        websocket_url = self._get_websocket_url(self._port)
+        logging.info("Connecting to websocket %s" % websocket_url)
+        ws = websocket.create_connection(
+            websocket_url, timeout=self._CONN_TIMEOUT
+        )
+        ws.settimeout(0)  # Don"t wait for new messages
+        return ws
 
-    def retry_if_exception(socket_handler_instance, *args, **kwargs):
+    @contextlib.contextmanager
+    def _ws_io(self):
 
+        # noinspection PyBroadException
         try:
-            return socket_handler_method(socket_handler_instance, *args, **kwargs)
+            yield
+        except Exception as e:
+            self.exception = e
+            logging.exception("WS messaging thread terminated with exception", exc_info=True)
+        finally:
+            self.close()
+
+    def close(self):
+        if hasattr(self, "ws") and self.ws:
+            try:
+                self.ws.close()
+            except websocket.WebSocketConnectionClosedException:
+                pass
+
+    def add_to_send_queue(self, payload):
+        self._send_queue.append(payload)
+
+    def get_from_recv_queue(self):
+        if self._recv_queue:
+            return self._recv_queue.pop(0)
+
+    def stop(self):
+        self._continue = False
 
-        except websocket.WebSocketConnectionClosedException:
+    def run(self):
 
-            socket_handler_instance.increment_connection_closed_count()
-            retry_if_exception(socket_handler_instance, *args, **kwargs)
+        with self._ws_io():
 
-    return retry_if_exception
+            self._last_poll = time.time()
+            while self._continue:
+
+                while self._send_queue:
+                    message = self._send_queue[0]
+                    self.ws.send(message)
+                    self._send_queue.pop(0)  # Don't pop first, in-case send excepts
+
+                while len(self._recv_queue) < self._MAX_QUEUE_BUFFER:
+                    try:
+                        message = self.ws.recv()
+                        self._recv_queue.append(message)
+                    except socket.error:
+                        break
```

Comment:
We don't care about `socket.error` here? If so, it would be nice to add a comment explaining why

Reply:
Hey @RodolfoSGomez we expect a socket.error("[Errno 11] Resource temporarily unavailable") when there are no more messages to read from the socket. I can make it a bit safer, and raise any exception that doesn't have that error message specifically.

Reply:
I'll also leave a comment explaining.

Reply:
Sure, sounds good